### ISSUE TITLE
Purging a queue while sending messages results in unacknowleged messages (Rabbit v3.6.2).

### ISSUE THREAD

**Issue by alaendle** [rabbitmq/rabbitmq-mqtt#84]
I've tested the following scenario - I just send a lot of messages (QoS 1) from multiple producers to a persistent/mirrored queue (without any consumers). If I now purge the contents of the queue via management UI some of the producers began to hang. If I record the network traffic via WireShark I can see that PUBACK's are missing - e.g. I send messages with identifiers 1 to 1000, but I receive only acknowledgements for 1 to 920 - so the client library is waiting for further acknowledgements, but rabbitmq won't send them - just leaving the tcp connection open. I think this violates the MQTT specification - http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718041. I know that I didn't give you much information, but maybe you already have a clue where to look at from my rough description of the problem. Also please don't hesitate to contact me if you need additional information or if I could otherwise help you to track down the problem. Thanks in advance for any hint!

**Comment by michaelklishin** [comment]
@alaendle what RabbitMQ version was used? What's in RabbitMQ log files? Can you post a code snippet that demonstrates the issue?

Purging a queue can affect consumers for a fairly obvious reason but not publishers.

**Comment by alaendle** [comment]
@michaelklishin I'm testing against v3.6.2. No lines in the log that I think are related to this problem. I use the paho java client to send messages via mqtt - here is the snippet I've used:

```
public static int asyncPublish(String broker, String topic, String userName, String password, String[] messages) {
        int qos = 1;                 // at least once
        String clientId = MqttClient.generateClientId() + new Random().nextInt(1000000);  // generate unique id to not interfere with other sessions;
        MqttAsyncClient client = null;
        try {
            client = new MqttAsyncClient(broker, clientId, new MemoryPersistence());
            MqttConnectOptions connOpts = new MqttConnectOptions();
            connOpts.setCleanSession(true);
            connOpts.setUserName(userName);
            connOpts.setPassword(password.toCharArray());
            connOpts.setMaxInflight(1000);
            connOpts.setKeepAliveInterval(0);
            log("Connecting to broker: " + broker);
            client.connect(connOpts).waitForCompletion();
            log("Connected");
            ArrayList deliveryTokens = new ArrayList();
            for (int i = 0; i < messages.length; i++) {
                log("Publishing message: " + messages[i]);
                CharToByteConverter converter = CharToByteConverter.getConverter("UTF-8");
                MqttMessage message = new MqttMessage(converter.convertAll(messages[i].toCharArray()));
                message.setQos(qos);
                deliveryTokens.add(client.publish(topic, message));
                log("Message published");
            }

            for(int i = 0; i < deliveryTokens.size(); i++) {
                ((IMqttDeliveryToken)deliveryTokens.get(i)).waitForCompletion();
            }

            client.disconnect().waitForCompletion();
            log("Disconnected");
            return 0;
        } catch (MqttException me) {
            log("reason " + me.getReasonCode());
            log("msg " + me.getMessage());
            log("loc " + me.getLocalizedMessage());
            log("cause " + me.getCause());
            log("excep " + me);
            me.printStackTrace();
            if(me.getCause()!=null) {
                me.getCause().printStackTrace();
            }
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        } catch (MalformedInputException e) {
            e.printStackTrace();
        } finally {
            if(client != null) {
                try {
                    client.close();
                } catch (MqttException e) {
                    e.printStackTrace();
                }
            }
        }

        return -1;
    }

public static void main(String[] args) throws InterruptedException {
        String[] msgs = new String[1000];
        Arrays.fill(msgs, "------------------PAYLOAD--------");
        for(int i = 0; i<1000; i++)
        {
            System.out.println(asyncPublish("tcp://***URL***", "***destination***", "***user***", "***pass***", msgs));
            Thread.sleep(1000);
        }
    }
```

Note that it seems that you need to run this code in parallel to have the "desired" effect of waiting forever in '((IMqttDeliveryToken)deliveryTokens.get(i)).waitForCompletion();'.

**Comment by michaelklishin** [comment]
Nonetheless, can you please post log files (including the SASL one)
somewhere around the time
of your test?

On Mon, Jun 13, 2016 at 4:56 PM, Andreas Ländle <notifications@github.com>
wrote:

> @michaelklishin <https://github.com/michaelklishin> I'm testing against
> v3.6.2. No lines in the log that I think are related to this problem. I use
> the paho java client to send messages via mqtt - here is the snippet I've
> used:
>
> public static int asyncPublish(String broker, String topic, String userName, String password, String[] messages) {
>         int qos = 1;                 // at least once
>         String clientId = MqttClient.generateClientId() + new Random().nextInt(1000000);  // generate unique id to not interfere with other sessions;
>         MqttAsyncClient client = null;
>         try {
>             client = new MqttAsyncClient(broker, clientId, new MemoryPersistence());
>             MqttConnectOptions connOpts = new MqttConnectOptions();
>             connOpts.setCleanSession(true);
>             connOpts.setUserName(userName);
>             connOpts.setPassword(password.toCharArray());
>             connOpts.setMaxInflight(1000);
>             connOpts.setKeepAliveInterval(0);
>             log("Connecting to broker: " + broker);
>             client.connect(connOpts).waitForCompletion();
>             log("Connected");
>             ArrayList deliveryTokens = new ArrayList();
>             for (int i = 0; i < messages.length; i++) {
>                 log("Publishing message: " + messages[i]);
>                 CharToByteConverter converter = CharToByteConverter.getConverter("UTF-8");
>                 MqttMessage message = new MqttMessage(converter.convertAll(messages[i].toCharArray()));
>                 message.setQos(qos);
>                 deliveryTokens.add(client.publish(topic, message));
>                 log("Message published");
>             }
>
>             for(int i = 0; i < deliveryTokens.size(); i++) {
>                 ((IMqttDeliveryToken)deliveryTokens.get(i)).waitForCompletion();
>             }
>
>             client.disconnect().waitForCompletion();
>             log("Disconnected");
>             return 0;
>         } catch (MqttException me) {
>             log("reason " + me.getReasonCode());
>             log("msg " + me.getMessage());
>             log("loc " + me.getLocalizedMessage());
>             log("cause " + me.getCause());
>             log("excep " + me);
>             me.printStackTrace();
>             if(me.getCause()!=null) {
>                 me.getCause().printStackTrace();
>             }
>         } catch (UnsupportedEncodingException e) {
>             e.printStackTrace();
>         } catch (MalformedInputException e) {
>             e.printStackTrace();
>         } finally {
>             if(client != null) {
>                 try {
>                     client.close();
>                 } catch (MqttException e) {
>                     e.printStackTrace();
>                 }
>             }
>         }
>
>         return -1;
>     }
>
> public static void main(String[] args) throws InterruptedException {
>         String[] msgs = new String[1000];
>         Arrays.fill(msgs, "------------------PAYLOAD--------");
>         for(int i = 0; i<1000; i++)
>         {
>             System.out.println(asyncPublish("tcp://***URL***", "***destination***", "***user***", "***pass***", msgs));
>             Thread.sleep(1000);
>         }
>     }
>
> Note that it seems that you need to run this code in parallel to have the
> "desired" effect of waiting forever in
> '((IMqttDeliveryToken)deliveryTokens.get(i)).waitForCompletion();'.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-mqtt/issues/84#issuecomment-225588085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAEQlTJT6rbaq8SR83EeemT87Vf4QA5ks5qLWGJgaJpZM4I0VE2>
> .
>



-- 
MK

Staff Software Engineer, Pivotal/RabbitMQ

**Comment by michaelklishin** [comment]
RabbitMQ should send `PUBACK` when messages are confirmed internally (the safest option), and if there really aren't any channel exceptions in the log then those messages were purged before they were persisted and it's likely that those won't be acknowledged.

There is no evidence to suggest that or something else but I'd question why this behaviour is wrong: if messages *actually* weren't handled by the server for one reason or another, why is it a good idea to send an ack to the client? Of course, what makes sense and what the spec says you should do isn't the same thing in messaging protocols.

**Comment by alaendle** [comment]
@michaelklishin Sorry for the late answer, I participated in an agile training today. I try to find time tomorrow to further investigate my code and the protocol specification. As it seems there are some discussion about NACK's in MQTT (https://issues.oasis-open.org/browse/MQTT-236), but for now they aren't part of the protocol :-( However I see drawbacks if rabbitMQ just doesn't send PUBACKS, because the only way a client can react is to wait (with timeout) and it is very likely that it tries to resubmit the messages in this case - and so the only visible effect of not acknowledging the message is to put the client, network and rabbit under stress ;-)

@jpwsutton Since you maintain the paho mqtt library and so I've expect you are very familiar with the protocol specification, maybe you can contribute to the discussion. And if I understand the current implementation the default of a synchronous publish would be to wait forever for an acknowledgement, right?

For sure I could implement a workaround but just setting a timeout, nonetheless I think it's an important topic for large scale systems since this effects throughput - and also might make it possible to block producer that expect a puback (like paho).

**Comment by michaelklishin** [comment]
@alaendle so assuming that RabbitMQ correctly doesn't send acks because messages were purged before persisted to a durable queue (QoS 1), we should workaround MQTT's lack of provisions that any real world system needs (NACKS, client error notifications, authorization) with hacks and incorrect behaviour on our end?

**Comment by alaendle** [comment]
@michaelklishin I agree this sounds somehow illogical. In the meanwhile I could reproduce this in a very simple scenario, just one node - no replication - just one producer, and just one bound queue (it just harder to hit the right time). So let me rephrase the problem a little bit.

I'm sending messages (to an exchange) - if I have no bound queue the messages are all acknowledged and are thrown away instantly. If a queue is bound, and what is done with it, should be transparent to the client - I hope we can agree an that? But exactly here is my problem: The purge of a queue should be transparent to the client, but in my case it is directly affecting the communication - no acks - to the mqtt client :-(

Do you still need some logfiles? Also I've noticed that in the web management console the blocked channel show's a Unconfirmed message. And I also tried this with RabbitMQ 3.7.0.607 (in conjunction with Erlang 18.3) and it is still reproducible.

**Comment by alaendle** [comment]
Might not be related to MQTT, I believe I can create basically the same dead-lock using the following C# code.

`using RabbitMQ.Client;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RabbitMQ.Client.Events;

namespace RabbitMQ
{
    class Program
    {
        const string QUEUE_NAME = "test";
        const int numberOfMessages = 1000000;

        public static void Main()
        {
            var factory = new ConnectionFactory() { HostName = "localhost:5673" };
            using (var connection = factory.CreateConnection())
            {
                var channel = connection.CreateModel();
                channel.ExchangeDeclare(QUEUE_NAME, ExchangeType.Fanout, true);
                channel.QueueDeclare(QUEUE_NAME, true, false, false, null);
                channel.QueueBind(QUEUE_NAME, QUEUE_NAME, String.Empty, new Dictionary<string, object>());
                channel.BasicAcks += channel_BasicAcks;
                channel.ConfirmSelect();

                for (var i = 1; i <= numberOfMessages; i++)
                {
                    var messageProperties = channel.CreateBasicProperties();
                    messageProperties.Persistent = true;

                    var message = String.Format("{0}\thello world", i);
                    var payload = Encoding.Unicode.GetBytes(message);
                    //Console.WriteLine("Sending message: " + message);
                    channel.BasicPublish(QUEUE_NAME, QUEUE_NAME, messageProperties, payload);

                    if(i%1000==0)channel.WaitForConfirmsOrDie();
                }
            }

            Console.WriteLine(" Press [enter] to exit.");
            Console.ReadLine();
        }

        private static void channel_BasicAcks(object sender, BasicAckEventArgs e)
        {
            Console.WriteLine(e.DeliveryTag + " - " + e.Multiple);
        }
    }
}
`

@michaelklishin How to proceed? Close this ticket and create one in the rabbitmq-server project?

**Comment by michaelklishin** [comment]
@alaendle post server logs to the mailing list. What you see as "a deadlock" could easily be a channel-level exception after which a channel cannot be used.

**Comment by alaendle** [comment]
@michaelklishin - There are absolutely no errors in the log, but just to make sure I've send the logs to the mailing list - https://groups.google.com/forum/#!topic/rabbitmq-users/9a_suP53Dq4
Again, many thanks for you help! Btw, in the meanwhile I work around the problem by just adding timeouts to all my asynchronous operations. However I think that we couldn't assume/guarantee that every client is also doing so - so a solution to this problem on the broker side might still be desirable.

**Comment by michaelklishin** [comment]
Let's continue on the list, if there are any findings we will open new issues where they belong.
